h/settings.py

```diff
@@ -13,6 +13,52 @@ class SettingError(Exception):
     pass
 
 
+class SettingsManager(object):
+    def __init__(self, settings, environ):
+        self.settings = settings
+        self.environ = environ
+
+    def set(self,
+            setting,
+            env_var,
+            type=str,
+            required=False,
+            default=None,
+            deprecated_msg=None):
+        """
+        Sets the specified setting in settings to the value of the env_var,
+        of if the env_var isn't present and a default value is specified; to
+        the default value..
+        If the env_var doesn't exist and the setting is not required then it
+        does nothing.
+        If the setting is required and not specified it raises a SettingError.
+        If a default value is provided and the env_var doesn't exist, the setting will be
+        overriden with the provided default value.
+        """
+        # If the environment var doesn't exist, there is no default set or provided,
+        # and the environment var is required, raise an error.
+        if (required and not default
+                and env_var not in self.environ
+                and setting not in self.settings):
+            raise SettingError(
+                'error parsing environment variable '
+                '{varname} not found'.format(varname=env_var))
```

Comment:
Can you try to re-write this code to avoid any compound `if` statements like this (three `and`s). I think the problem you've got into is by trying to handle the case of a missing required setting first. If you handle the cases in the opposite order I think it might look much simpler. Would something like this work:

```python
def set(...):

    def maybe_warn():
        if deprecated:
            log.warn(...)

    if envvar in environ:
        settings[name] = environ[envvar]
        maybe_warn()
    elif name in settings:
        maybe_warn()
    elif default:
        settings[name] = default
    elif required:
        raise SettingError()
```